### PR TITLE
Allow cross-compiling doctests

### PR DESCRIPTION
**Pull request by Goirad**
This PR allows doctest to receive a --runtool argument, as well as possibly many --runtool-arg arguments, which are then used to run cross compiled doctests.
Also, functionality has been added to rustdoc to allow it to skip testing doctests on a per-target basis, in the same way that compiletest does it. For example, tagging the doctest with "ignore-sgx" disables testing on any targets that contain "sgx". A plain "ignore" still skips testing on all targets.

See [here](https://github.com/rust-lang/cargo/pull/6892) for the companion PR in the cargo project that extends functionality in Cargo so that it passes the appropriate parameters to rustdoc when cross compiling and testing doctests.

Part of [#6460](https://github.com/rust-lang/cargo/issues/6460)

### CODE REVIEW

File: src/bootstrap/bootstrap.py [hunk 1]
```diff
@@ -678,7 +678,7 @@ def update_submodule(self, module, checked_out, recorded_submodules):
         run(["git", "submodule", "-q", "sync", module],
             cwd=self.rust_root, verbose=self.verbose)
         run(["git", "submodule", "update",
-            "--init", "--recursive", "--progress", module],
```
Comment (jethrogb) [inline src/bootstrap/bootstrap.py]: Why is this change needed?

File: src/librustdoc/html/markdown.rs [hunk 2]
```diff
@@ -645,7 +653,14 @@ impl LangString {
                     seen_rust_tags = seen_other_tags == false;
                 }
                 "no_run" => { data.no_run = true; seen_rust_tags = !seen_other_tags; }
-                "ignore" => { data.ignore = true; seen_rust_tags = !seen_other_tags; }
+                "ignore" => {
+                    data.ignore = Ignore::All;
+                    seen_rust_tags = !seen_other_tags;
+                }
+                x if x.starts_with("ignore-") => {
```
Comment (jethrogb) [inline src/librustdoc/html/markdown.rs]: Is this insta-stable?

File: src/bootstrap/bootstrap.py (hunk 1 above)
Comment (froydnj) [reply src/bootstrap/bootstrap.py]: I think this is to work around #57080.  I have an open request in #60379 to try and fix that issue.

Comment (Goirad) [reply src/bootstrap/bootstrap.py]: That is exactly correct, I'm going to revert it on the next commit

File: src/librustdoc/html/markdown.rs (hunk 2 above)
Comment (Goirad) [reply src/librustdoc/html/markdown.rs]: I do not know what that means

Comment (jethrogb) [reply src/librustdoc/html/markdown.rs]: I think you're adding a feature to the language that is immediately available to users without a feature gate.

Comment (Goirad) [reply src/librustdoc/html/markdown.rs]: Oh, I think that is how I wrote it. I didn't even consider having the new behavior as an option.

**Comment by Goirad** [comment]
Awaiting review, all checks pass now.

**Comment by jethrogb** [comment]
`compiletest` does space-splitting on `runtool`, but `rustdoc` does not, currently. So if your runtool contains a space you'll get different behavior for `rustdoc` and `compiletest`.

**Comment by Goirad** [comment]
@kennytm I have restricted the scope of this PR to only adding runtool functionality to rustdoc, along with Cargo in the other PR.

File: src/librustdoc/lib.rs [hunk 3]
```diff
@@ -354,6 +354,18 @@ fn opts() -> Vec<RustcOptGroup> {
                       "show-coverage",
                       "calculate percentage of public items with documentation")
         }),
+        stable("runtool", |o| {
```
Comment (kennytm) [inline src/librustdoc/lib.rs]: These two should be `unstable`

File: src/librustdoc/test.rs [hunk 4]
```diff
@@ -362,7 +370,17 @@ fn run_test(
...
     // Run the code!
-    let mut cmd = Command::new(output_file);
+    let mut cmd;
+
+    if let Some(tool) = runtool {
+        cmd = Command::new(tool);
+        cmd.arg(output_file);
+        for arg in runtool_args {
+            cmd.arg(arg);
+        }
```
Comment (kennytm) [inline src/librustdoc/test.rs]: Use `cmd.args(runtool_args)` here.

**Review by kennytm — COMMENTED** [review]
Are those `ignore` changes still in the scope of this PR? They're still in the diff.

**Comment by Goirad** [comment]
Yes ignore is still a part of this PR, since being able to ignore on a per-target basis is relevant to using the crosscompilation feature.

**Comment by kennytm** [comment]
🤔 It's better to delegate the review to @rust-lang/rustdoc then, as it is changing the behavior of ````` ```rust,ignore-foo `````.

**Comment by Goirad** [comment]
Sorry, how would that behavior change?

**Comment by kennytm** [comment]
@Goirad yeah that's what the "insta-stable" comment in https://github.com/rust-lang/rust/pull/60387#discussion_r279578483 is about.

File: src/librustdoc/html/markdown.rs (hunk 2 above)
Comment (jethrogb) [reply src/librustdoc/html/markdown.rs]: The last similar change was #54350 which looks like it just landed without stability concerns.

**Comment by GuillaumeGomez** [comment]
Just like others told you, please put it behind a feature gate. Code looks good otherwise, just fix the nit issues and we should be almost good to go.

**Comment by Goirad** [comment]
@GuillaumeGomez I have added the feature gate and fixed the nits

File: src/librustdoc/test.rs [hunk 5]
```diff
@@ -362,7 +371,15 @@ fn run_test(
...
 
     // Run the code!
-    let mut cmd = Command::new(output_file);
+    let mut cmd;
+
+    if let Some(tool) = runtool {
+        cmd = Command::new(tool);
+        cmd.arg(output_file);
+        cmd.args(runtool_args);
+    }else{
```
Comment (GuillaumeGomez) [inline src/librustdoc/test.rs]: Please update to `} else {`.

**Comment by GuillaumeGomez** [comment]
Ok, just one nit and then it's good for me.

Can you take a look as well @ollie27 please? :)

File: src/librustdoc/test.rs (hunk 5 above)
Comment (Goirad) [reply src/librustdoc/test.rs]: fixed

File: src/librustdoc/html/markdown.rs [hunk 6]
```diff
@@ -631,7 +640,11 @@ impl LangString {
                     seen_rust_tags = seen_other_tags == false;
                 }
                 "no_run" => { data.no_run = true; seen_rust_tags = !seen_other_tags; }
-                "ignore" => { data.ignore = true; seen_rust_tags = !seen_other_tags; }
+                "ignore" => { data.ignore = Ignore::All; seen_rust_tags = !seen_other_tags; }
+                x if enable_per_target_ignores && x.starts_with("ignore-") => {
```
Comment (QuietMisdreavus) [inline src/librustdoc/html/markdown.rs]: I'm concerned that this forces people to send `--enable-per-target-ignores` even when building docs regularly. Rustdoc uses the `LangString` struct even during docs generation to see whether it should syntax-highlight a doctest, and also to display the little banners that show on `no_run` and `ignore` doctests. This means that the way this is written right now, adding `ignore-something` to a doctest and *not* sending `--enable-per-target-ignores` will set `seen_other_tags` below, and cause a doctest to show up as plain-text. One way to fix this would be to always handle `ignore-something` lang strings, and ignore them if `--enable-per-test-ignores` isn't set:

```suggestion
                x if x.starts_with("ignore-") => if enable_per_target_ignores {
```

File: src/librustdoc/test.rs [hunk 7]
```diff
@@ -736,12 +762,22 @@ impl Tester for Collector {
...
         debug!("Creating test {}: {}", name, test);
         self.tests.push(testing::TestDescAndFn {
             desc: testing::TestDesc {
                 name: testing::DynTestName(name.clone()),
-                ignore: config.ignore,
+                ignore: match config.ignore {
+                    Ignore::All => true,
+                    Ignore::None => false,
+                    Ignore::Some(ref ignores) => {
+                        ignores.iter().any(|s| target_str.contains(s))
```
Comment (QuietMisdreavus) [inline src/librustdoc/test.rs]: Is there a better method of testing the `ignore-` string than this? I was hoping we could reuse [compiletest's filters](https://rust-lang.github.io/rustc-guide/tests/adding.html#ignoring-tests), which don't seem like they could work with a substring search.

**Review by QuietMisdreavus — COMMENTED** [review]
Thanks so much for writing this up! I have a couple comments, but it *seems* like the basic principle is sound. I may need to come back later on and give another read to be 100% sure of the complete implementation. Regardless, this is big! I'm excited to see this working.

Regarding bikeshedding the `ignore-` filters: One consequence of changing how we parse the lang strings like this is that when people start using new lang string options, older versions of rustdoc may no longer recognize those doctests, since they'll show up as an unknown lang string. However, in this case, if you write `rust,ignore-something` instead, then current versions of rustdoc will still see (and try to execute) the doctest. It may be worth allowing `ignore-something` to override `ignore` so that crates can still pass tests on older versions of rustdoc while the feature is still new. The lang string `rust,ignore,ignore-something` is a little verbose and seemingly contradictory, but hopefully once the feature lands it'll only be necessary as long as stable doesn't recognize it.

One thing i'm curious about is whether we should also support *only* running a specific doctest on a given filter, like compiletest's `only-` filters. I can imagine wanting to display some platform-specific behavior in a doctest, and adding `ignore-linux,ignore-windows,ignore-bsd` to your test to ensure it only runs on macOS (for example) quickly becomes way too cumbersome to properly ensure.

I'm also amused about how the `ignore-` feature overlaps somewhat with [`#[doc(cfg)]`](https://doc.rust-lang.org/nightly/rustdoc/unstable-features.html#documenting-platform-feature-specific-information). That feature already restricts running doctests unless the target matches the given cfg string. In some sense, this is even better, since you can filter individual doctests, but (as of the current state of the PR) there's no visibility on the generated docs about where a doctest is intended to run. (With `#[doc(cfg)]`, there's a banner printed on the item in question stating "this item is only available on Windows" or the like.)

It would be nice to see tests that exercise that doctests are run (or not run) on certain platforms. That may have to be a collection of tests in `src/test/rustdoc-ui` that filter the platform they run on and have some combination of flags that let it run the doctest (running a 32-bit doctest on its 64-bit counterpart and asserting that the bit width is correct, maybe?) combined with other tests that ensure that a doctest isn't run on the wrong platform (by ensuring that the console output states that the test was ignored).

I would also like to see some documentation in the Rustdoc Book (`src/doc/rustdoc/src/unstable-features.md`) that mention what these flags do. This is a fairly major feature, and relying on word-of-mouth or external documentation is less-than-ideal in making sure people try it out.

File: src/librustdoc/test.rs [hunk 8]
```diff
@@ -59,7 +60,7 @@ pub fn run(options: Options) -> i32 {
         edition: options.edition,
         ..config::Options::default()
     };
-
+    options.target.as_ref().map(|t| { sessopts.target_triple = t.clone() });
```
[the remark this answers is not in the document]
Comment (Goirad) [reply src/librustdoc/test.rs]: That's fair, and now that you mention it does look a little off, I'll change it.

File: src/librustdoc/config.rs [hunk 9]
```diff
@@ -476,6 +487,9 @@ impl Options {
         let generate_search_filter = !matches.opt_present("disable-per-crate-search");
         let persist_doctests = matches.opt_str("persist-doctests").map(PathBuf::from);
         let generate_redirect_pages = matches.opt_present("generate-redirect-pages");
+        let runtool = matches.opt_str("runtool");
+        let runtool_args = matches.opt_strs("runtool-arg");
```
[the remark this answers is not in the document]
Comment (Goirad) [reply src/librustdoc/config.rs]: I'm a bit confused by your first paragraph, it is my understanding that `opt_strs` returns a vec of all arguments passed to (possibly many) `--runtool-arg`'s. The intention being that arguments get passed one at a time, one per `--runtool-arg`. 

As for the question of integration with bootstrap, I am working on that now, but wanted to get the ball rolling here in rustdoc. My plan is to add an unstable flag to bootstrap that makes it pass `--Zdoctest-xcompile` to cargo test invocations. The current state is that compiletests do get cross-compiled, and it already has a `--runtool` option. My change to compiletest removes that option and instead uses Cargo to create the merged `.cargo/config` for a directory and use the runner specified there, with the idea being that it will ensure consistent testing across rustdoc and compiltest.

File: src/librustdoc/test.rs [hunk 10]
```diff
@@ -736,12 +763,23 @@ impl Tester for Collector {
...
         self.tests.push(testing::TestDescAndFn {
             desc: testing::TestDesc {
                 name: testing::DynTestName(name.clone()),
-                ignore: config.ignore,
+                ignore: match config.ignore {
+                    Ignore::All => true,
+                    Ignore::None => false,
+                    Ignore::Some(ref ignores) => {
+                        target_str.map_or(false,
+                                          |s| ignores.iter().any(|t| s.contains(t)))
```
[the remark this answers is not in the document]
Comment (Goirad) [reply src/librustdoc/test.rs]: Outdated, answered below

File: src/librustdoc/test.rs (hunk 7 above)
Comment (Goirad) [reply src/librustdoc/test.rs]: I have not investigated leveraging any existing machinery from compiltest's filters, although in the worst case it would not be too much trouble to extend the flags to handle only and some other target details.

File: src/librustdoc/html/markdown.rs (hunk 6 above)
Comment (Goirad) [reply src/librustdoc/html/markdown.rs]: This is a fair point, and your change would prevent that problem. To also bring in something you mentioned in your main comment, right now rustdoc does not test anything with an invalid filter. So running `rustdoc --test` with an older version of rustdoc has the effect of silently ignoring any doctests with unknown (ie newer) filters.

File: src/librustdoc/test.rs (hunk 7 above)
Comment (jethrogb) [reply src/librustdoc/test.rs]: I don't think the compiletest filters are particularly well-suited to stabilizing. They just define a custom list of keywords for matching platforms. Since compiletest is for internal use and this is not, I think we can put a little more thought into this.

I do agree that substring matching on the target name is a little strange. It would be nice to somehow hook into Rust's existing conditional compilation story, which is `cfg`.

**Comment by Goirad** [comment]
First, thanks for the review @QuietMisdreavus ! I really appreciate the input. To address the rest of your comments, I can definitely add some tests, at least for the functionality I have added, as well as update the book.
As far as allowing `ignore-something` to override `ignore`, that seems like a fine idea, and fairly simple to implement, I can add that, assuming we don't leave this conversation with a substantial rework planned.

File: src/librustdoc/config.rs (hunk 9 above)
[the remark this answers is not in the document]
Comment (Goirad) [reply src/librustdoc/config.rs]: I see now what you meant in your first paragraph. You're right it would be more robust, I just omitted any error checking there, since worst case scenario one string that is in reality several instructions would just get passed wholesale to the runtool (which I think would not in the end be any different from adding them one at a time). 
Also as far as who to ask, I don't know, but I welcome any suggestions since this seems to be pretty central to how the whole rust compiler building and testing process works.

File: src/librustdoc/test.rs (hunk 7 above)
Comment (Goirad) [reply src/librustdoc/test.rs]: In the end I think this is sufficient for an initial implementation of an unstable feature, and others can expand this as the need arises.

File: src/librustdoc/config.rs (hunk 9 above)
[the remark this answers is not in the document]
Comment (QuietMisdreavus) [reply src/librustdoc/config.rs]: > (which I think would not in the end be any different from adding them one at a time)

Since we would be building up a Command instance with these args, the space-splitting wouldn't happen - the recipient would receive the blob in one "argument", as if it had all been quoted together.

One option would be to look at how many `--runtool-arg` arguments there are and treat it differently based on whether there are one or more than one. If there's only one argument, we could do space-splitting on that, and if there is more than one argument, we can leave them as-is. This somewhat mirrors [Cargo's own parsing of the relevant configuration item](https://github.com/rust-lang/cargo/blob/807429e1b6da4e2ec52488ef2f59e77068c31e1f/src/cargo/util/config.rs#L509-L529), which whitespace-splits it is it's just a string, and leaves them alone if given a list. (This can backfire if someone tried to give a single argument with spaces in it, though... :thinking:)

[the remark this answers is not in the document]
Comment (jethrogb) [reply src/librustdoc/config.rs]: It's not possible to correctly space-split command-line arguments based on some heuristic if you also want to support arguments with spaces in them. Just let the `cargo`/the user take care of this. Most people won't be running `rustdoc` directly this way anyway.

**Comment by Goirad** [comment]
@ollie27 review ping

**Comment by Goirad** [comment]
@ollie27 @QuietMisdreavus @GuillaumeGomez Can I get some review please?

**Comment by JohnCSimon** [comment]
ping from triage - 
@Goirad  @ollie27 @QuietMisdreavus @GuillaumeGomez 
This PR has a merge conflict, but as soon as it's resolved it's ready for review. 
Thank you for your contributions!

> @ollie27 @QuietMisdreavus @GuillaumeGomez Can I get some review please?

**Comment by JohnTitor** [comment]
Ping from triage: @Goirad there is a conflict to resolve, and @ollie27 could you review this?

**Comment by GuillaumeGomez** [comment]
Just to be sure: shouldn't it be done over https://github.com/rust-lang/rust/pull/63827 considering they'll make doctests run outside of rustdoc?

**Comment by jethrogb** [comment]
This PR has been open for months, if @ollie27 is not available for review, can the docs team appoint a different reviewer?

**Comment by ollie27** [comment]
Sorry for the delay.

As far as I'm concerned this feature and code looks good so r=me with a rebase.

This could do with some tests like `run-make` for these new features but that can be in a follow up PR.

We also need tracking issues for the new features.

**Comment by Goirad** [comment]
@ollie27 I have opened a tracking issue [here](https://github.com/rust-lang/rust/issues/64245).

Thanks for the review!

**Comment by ollie27** [comment]
Thanks, lets get this in.

@bors r+
